### PR TITLE
8339317: Optimize ClassFile writeBuffer

### PR DESCRIPTION
**Pull request by wenshao**
A small optimization, optimize the BufferWriter implementation and use of ClassFile, provide faster patchInt and skip

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 2 [Reviewers](https://openjdk.org/bylaws#reviewer))

### Issue
 * [JDK-8339317](https://bugs.openjdk.org/browse/JDK-8339317): Optimize ClassFile writeBuffer (**Enhancement** - P4)

### Reviewers
 * [Claes Redestad](https://openjdk.org/census#redestad) (@cl4es - **Reviewer**)
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20780/head:pull/20780` \
`$ git checkout pull/20780`

Update a local copy of the PR: \
`$ git checkout pull/20780` \
`$ git pull https://git.openjdk.org/jdk.git pull/20780/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20780`

View PR using the GUI difftool: \
`$ git pr show -t 20780`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20780.diff">https://git.openjdk.org/jdk/pull/20780.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20780#issuecomment-2321061943)

### CODE REVIEW

**Comment by liach** [comment]
`patchInt` isn't that frequently used in workloads. I doubt this specialized version provides much value; maybe @cl4es can evaluate?

**Comment by wenshao** [comment]
I debugged the code and watched the compile log, patchInt calls are frequent. For example, the following call stack:

```
patchInt:164, BufWriterImpl (jdk.internal.classfile.impl)
writeTo:856, UnboundAttribute$AdHocAttribute (jdk.internal.classfile.impl)
writeAttribute:233, Util (jdk.internal.classfile.impl)
writeAttributes:240, Util (jdk.internal.classfile.impl)
writeTo:53, AttributeHolder (jdk.internal.classfile.impl)
writeTo:153, DirectMethodBuilder (jdk.internal.classfile.impl)
writeList:247, Util (jdk.internal.classfile.impl)
build:187, DirectClassBuilder (jdk.internal.classfile.impl)
build:125, ClassFileImpl (jdk.internal.classfile.impl)
build:332, ClassFile (java.lang.classfile)
classFileSetup:261, InvokerBytecodeGenerator (java.lang.invoke)
generateCustomizedCodeBytes:540, InvokerBytecodeGenerator (java.lang.invoke)
generateCustomizedCode:518, InvokerBytecodeGenerator (java.lang.invoke)
compileToBytecode:845, LambdaForm (java.lang.invoke)
makePreparedLambdaForm:302, DirectMethodHandle (java.lang.invoke)
preparedLambdaForm:230, DirectMethodHandle (java.lang.invoke)
preparedLambdaForm:215, DirectMethodHandle (java.lang.invoke)
preparedLambdaForm:224, DirectMethodHandle (java.lang.invoke)
make:106, DirectMethodHandle (java.lang.invoke)
make:127, DirectMethodHandle (java.lang.invoke)
make:132, DirectMethodHandle (java.lang.invoke)
basicInvoker:96, Invokers (java.lang.invoke)
<init>:1102, LambdaForm$NamedFunction (java.lang.invoke)
<init>:1354, LambdaForm$Name (java.lang.invoke)
makeReinvokerForm:160, DelegatingMethodHandle (java.lang.invoke)
makeReinvokerForm:120, DelegatingMethodHandle (java.lang.invoke)
chooseDelegatingForm:112, DelegatingMethodHandle (java.lang.invoke)
<init>:50, DelegatingMethodHandle (java.lang.invoke)
<init>:465, MethodHandleImpl$AsVarargsCollector (java.lang.invoke)
<init>:462, MethodHandleImpl$AsVarargsCollector (java.lang.invoke)
makeVarargsCollector:453, MethodHandleImpl (java.lang.invoke)
asVarargsCollector:1516, MethodHandle (java.lang.invoke)
withVarargs:1201, MethodHandle (java.lang.invoke)
setVarargs:1720, MethodHandle (java.lang.invoke)
getDirectMethodCommon:4122, MethodHandles$Lookup (java.lang.invoke)
getDirectMethodNoSecurityManager:4065, MethodHandles$Lookup (java.lang.invoke)
getDirectMethodForConstant:4314, MethodHandles$Lookup (java.lang.invoke)
linkMethodHandleConstant:4262, MethodHandles$Lookup (java.lang.invoke)
linkMethodHandleConstant:628, MethodHandleNatives (java.lang.invoke)
concat6String:167, StringConcat (jmh)
concat6String:174, StringConcatTest (jmh)
main:239, StringConcatTest (jmh)
```

File: src/java.base/share/classes/jdk/internal/classfile/impl/DirectCodeBuilder.java [hunk 1]
```diff
@@ -469,8 +469,12 @@ private void processDeferredLabels() {
         if (deferredLabels != null) {
             for (DeferredLabel dl : deferredLabels) {
                 int branchOffset = labelToBci(dl.label) - dl.instructionPc;
-                if (dl.size == 2 && (short)branchOffset != branchOffset) throw new LabelOverflowException();
-                bytecodesBufWriter.patchInt(dl.labelPc, dl.size, branchOffset);
+                if (dl.size == 2) {
+                    if ((short)branchOffset != branchOffset) throw new LabelOverflowException();
+                    bytecodesBufWriter.patchU2(dl.labelPc, branchOffset);
+                } else {
+                    bytecodesBufWriter.patchInt(dl.labelPc, branchOffset);
```
Comment (cl4es) [inline src/java.base/share/classes/jdk/internal/classfile/impl/DirectCodeBuilder.java:476]: Should assert or test that `dl.size == 4` here.

File: src/java.base/share/classes/jdk/internal/classfile/impl/AbstractAttributeMapper.java [hunk 2]
```diff
@@ -62,13 +62,14 @@ public final String name() {
     }
 
     @Override
-    public final void writeAttribute(BufWriter buf, T attr) {
+    public final void writeAttribute(BufWriter writer, T attr) {
+        BufWriterImpl buf = (BufWriterImpl) writer;
         buf.writeIndex(buf.constantPool().utf8Entry(name));
         buf.writeInt(0);
```
Comment (cl4es) [inline src/java.base/share/classes/jdk/internal/classfile/impl/AbstractAttributeMapper.java:68]: Since this is the int we're patching later it might be reasonable to have a buf.skip(int) method which just increases buf.offset. Applicable at least here and in `UnboundAttribute::writeTo`

File: src/java.base/share/classes/jdk/internal/classfile/impl/Util.java [hunk 3]
```diff
@@ -281,8 +281,8 @@ public void writeBody(BufWriterImpl b) {
                                     b.writeU2(-1);//max locals
                                     b.writeInt(bytecode.limit());
                                     b.writeBytes(bytecode.array(), 0, bytecode.limit());
-                                    b.writeU2(0);//exception handlers
-                                    b.writeU2(0);//attributes
+                                    b.skip(2);//exception handlers
+                                    b.skip(2);//attributes
```
Comment (cl4es) [inline src/java.base/share/classes/jdk/internal/classfile/impl/Util.java:285]:
```suggestion
                                    b.skip(4); //exception handlers (2) + attributes (2)
```

File: src/java.base/share/classes/jdk/internal/classfile/impl/BufWriterImpl.java [hunk 4]
```diff
@@ -175,6 +189,14 @@ public void writeIntBytes(int intSize, long intValue) {
         }
     }
 
+    public void skip(int skipSize) {
+        int nextOffset = offset + skipSize;
+        if (nextOffset > elems.length) {
+            grow(nextOffset);
+        }
+        this.offset = nextOffset;
```
Comment (cl4es) [inline src/java.base/share/classes/jdk/internal/classfile/impl/BufWriterImpl.java:197]: Let's not duplicate code to get rid of a simple addition
```suggestion
        reserveSpace(skipSize);
        this.offset += skipSize;
```

**Review by cl4es — COMMENTED** [review]
Looks ok, a couple of nits.

**Review by cl4es — APPROVED** [review]

File: src/java.base/share/classes/jdk/internal/classfile/impl/BufWriterImpl.java [hunk 5]
```diff
@@ -175,6 +189,11 @@ public void writeIntBytes(int intSize, long intValue) {
         }
     }
 
+    public void skip(int skipSize) {
```
Comment (liach) [inline src/java.base/share/classes/jdk/internal/classfile/impl/BufWriterImpl.java:192]: Conceptually I would prefer to make such an API return an int for skipped index so we can more easily patch later. But that is really optional.

File: src/java.base/share/classes/jdk/internal/classfile/impl/BufWriterImpl.java [hunk 6]
```diff
@@ -218,7 +237,7 @@ public void writeIndex(PoolEntry entry) {
     @Override
     public void writeIndexOrZero(PoolEntry entry) {
         if (entry == null || entry.index() == 0)
-            writeU2(0);
+            skip(2);
```
Comment (liach) [inline src/java.base/share/classes/jdk/internal/classfile/impl/BufWriterImpl.java:240]: Writing `0` CP index for `null` is mandated by the attributes that do so. https://docs.oracle.com/javase/specs/jvms/se22/html/jvms-4.html#jvms-4.7.24-300-D-A.1

File: src/java.base/share/classes/jdk/internal/classfile/impl/DirectCodeBuilder.java [hunk 7]
```diff
@@ -590,13 +595,13 @@ public void writeInvokeInterface(Opcode opcode,
         writeBytecode(opcode);
         bytecodesBufWriter.writeIndex(ref);
         bytecodesBufWriter.writeU1(count);
-        bytecodesBufWriter.writeU1(0);
+        bytecodesBufWriter.skip(1);
```
Comment (liach) [inline src/java.base/share/classes/jdk/internal/classfile/impl/DirectCodeBuilder.java:598]: `0` is mandated: https://docs.oracle.com/javase/specs/jvms/se22/html/jvms-6.html#jvms-6.5.invokeinterface

File: src/java.base/share/classes/jdk/internal/classfile/impl/DirectCodeBuilder.java [hunk 8]
```diff
@@ -590,13 +595,13 @@ public void writeInvokeInterface(Opcode opcode,
...
     }
 
     public void writeInvokeDynamic(InvokeDynamicEntry ref) {
         writeBytecode(INVOKEDYNAMIC);
         bytecodesBufWriter.writeIndex(ref);
-        bytecodesBufWriter.writeU2(0);
+        bytecodesBufWriter.skip(2);
```
Comment (liach) [inline src/java.base/share/classes/jdk/internal/classfile/impl/DirectCodeBuilder.java:604]: `0` is mandated: https://docs.oracle.com/javase/specs/jvms/se22/html/jvms-6.html#jvms-6.5.invokedynamic

**Review by liach — CHANGES_REQUESTED** [review]
This migration of write 0 to skip reveals that they are different. Good RFE. (Maybe we can make the BufWriterImpl buffer growth to allocate non-initialized array after the write 0 fixes)

File: src/java.base/share/classes/jdk/internal/classfile/impl/AbstractPoolEntry.java [hunk 9]
```diff
@@ -446,7 +446,7 @@ void writeTo(BufWriterImpl pool) {
                             throw new IllegalArgumentException();
                         }
                         int byteLengthFinal = byteLength;
-                        pool.patchInt(pool.size() - i - 2, 2, byteLengthFinal);
+                        pool.patchU2(pool.size() - i - 2, byteLengthFinal);
```
Comment (liach) [inline src/java.base/share/classes/jdk/internal/classfile/impl/AbstractPoolEntry.java:449]: This change is causing conflict with your utf8 writing enhancement. Can you revert this change and squash commits, or merge latest master to fix this conflict?

**Review by liach — APPROVED** [review]

**Review by cl4es — APPROVED** [review]

**Review by liach — APPROVED** [review]
